refactor: add base path in Operate and Tasklist apps urls

houssain-barouni opened the pull request:
## Description
The Camunda single application should serve Zeebe, Operate and Tasklist using one running application.

To dispatch between Operate and Tasklist UIs, we need to add `/operate` and `/tasklist` base paths to the URLs respectively used by Operate and Tasklist UIs
For consistency, the same base paths will be used when Operate and Tasklist are run in standalone mode.

- requests to `/tasklist` and `/operate` are redirected to Tasklist or Operate index file
- requests to the root domain are redirect to either `/tasklist` or `/operate` depending on the used profile in standalone mode, it defaults to `/operate` in single application mode.
- backend API urls do not change and do not contain the base paths.
- old app URLs not containing the base path (example `http://localhost:8080/processes`) will not be working, the app base path should be provided in the URL (example `http://localhost:8080/tasklist/processes`) 
- Tasklist public forms URLs (`/new/{bpmnProcessId}`do not require the base path and will continue to be working as they are served by a [dedicated endpoint in backend](https://github.com/camunda/camunda/blob/main/tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/api/rest/v1/controllers/PublicProcessController.java).


## Related issues

relates to https://github.com/camunda/zeebe/issues/18535

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java:
```diff
@@ -0,0 +1,125 @@
+    }
+  }
+
+  private boolean isStandaloneTasklist() {
+    return environment.acceptsProfiles(TASKLIST.getId(), "!" + OPERATE.getId());
+  }
```
We must get rid of such code patterns. That means we should not introduce a pattern in which we are checking the profile everywhere and, depending on it, it behaves differently. That way, we will lose the overview.

Instead, we should configure the application and make sure that the code behaves the same way.

houssain-barouni replied:
Done! new property `camunda.webapps.default-app` is introduced and set by `WebappsConfigurationInitializer.java`

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java:
```diff
@@ -0,0 +1,125 @@
+    } else if (isLoginDelegated() && isNotLoggedIn()) {
+      return saveRequestAndRedirectToLogin(request, requestedURL);
+    } else {
+      if (isStandaloneTasklist()) {
+        return "forward:/tasklist";
+      } else {
+        return "forward:/operate";
+      }
```
In this case, the code can assume that the request URL starts with `tasklist` or `operate`, so it could use it and redirect to what is set in the URL (i.e., with what the URL starts). That way, the additional check if it is a standalone tasklist or not is not necessary.

houssain-barouni replied:
yes it makes sense 
besides that the forwarding in the last else block is not correct, we should forward to the requested app and the default one :)

houssain-barouni replied:
fixed!

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  @RequestMapping(value = {"/{regex:[\\w-]+}", "/**/{regex:[\\w-]+}"})
+  public String forward404(final HttpServletRequest request, final HttpServletResponse response) {
```
To me, this error controller seems to mix the exclude and include patterns. What cases must it handle? Can we implement it so that it only catches the necessary cases (i.e., only the included cases)?

houssain-barouni replied:
I will check if there is a better way

houssain-barouni replied:
Finally no need for the exclude pattern, `ForwardErrorController` will only handle `/operate/**` and `/tasklist/**` patterns

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java:
```diff
@@ -0,0 +1,125 @@
+        || !authentication.isAuthenticated();
+  }
+
+  private boolean isLoginDelegated() {
+    return Arrays.stream(environment.getActiveProfiles())
+        .anyMatch(Set.of(IDENTITY_AUTH.getId(), SSO_AUTH.getId())::contains);
+  }
```
Again, it should be a configuration and not depending on the profile.

houssain-barouni replied:
`isLoginDelegated` is a legacy method, but we can sort it out

houssain-barouni replied:
Done! new property `camunda.webapps.loginDeletgated` is introduced and set by WebappsConfigurationInitializer.java

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/IndexController.java:
```diff
@@ -5,22 +5,40 @@
-    model.addAttribute("contextPath", context.getContextPath() + "/");
-    return "index";
+  public String index() {
+    if (environment.acceptsProfiles(TASKLIST.getId(), "!" + OPERATE.getId())) {
```
The same as above, it should be configured what is the default app, so that it redirects to it. It should not know anything about profiles at this place.

houssain-barouni replied:
Done! new property `camunda.webapps.default-app` is introduced and set by `WebappsConfigurationInitializer`

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/IndexController.java:
```diff
@@ -5,22 +5,40 @@
-package io.camunda.operate.webapp;
+package io.camunda.webapps.controllers;
+
+import static io.camunda.application.Profile.*;
```
I think no rule enforces it, but can you ensure that it lists the used one, and not only `*`

houssain-barouni replied:
`import static io.camunda.application.Profile.*` should be deleted anyway

romansmirnov commented on dist/src/main/java/io/camunda/application/StandaloneTasklist.java:
```diff
@@ -46,6 +48,7 @@
           pattern = "io\\.camunda\\.tasklist\\.archiver\\..*")
     },
     nameGenerator = FullyQualifiedAnnotationBeanNameGenerator.class)
+@Import(WebappsModuleConfiguration.class)
```
I suggest to migrate the `StandaloneApplication` to use the `MainSupport` to construct an application as done in `StandaloneOperate`. And the `WebappsModuleConfiguration` should be integrated as an addition `source`.

houssain-barouni replied:
I was planning to make `StandaloneTasklist` use the `MainSupport` in a next PR, but I will do it here if it makes it cleaner
👍  for adding `WebappsModuleConfiguration` in  `StandaloneOperate` and `StandaloneCamunda`

romansmirnov commented on dist/src/main/java/io/camunda/operate/OperateModuleConfiguration.java:
```diff
@@ -39,6 +41,7 @@
     // use fully qualified names as bean name, as we have classes with same names for different
     // versions of importer
     nameGenerator = FullyQualifiedAnnotationBeanNameGenerator.class)
+@Import(WebappsModuleConfiguration.class)
```
As suggested above, let's remove it.

romansmirnov commented on dist/src/main/java/io/camunda/webapps/WebappsModuleConfiguration.java:
```diff
@@ -0,0 +1,18 @@
+@Configuration(proxyBeanMethods = false)
+@ComponentScan(basePackages = "io.camunda.webapps")
+@EnableAutoConfiguration
+public class WebappsModuleConfiguration {}
```
Let's depend this on the profiles `tasklist` or `operate` as done with the app-specific configuration classes.

houssain-barouni replied:
Done! `WebappsModuleConfiguration` is added in the sources used in StandaloneOperate, StandaloneTasklist and StandaloneCamunda. It is also enabled with `camunda.webapps.enabled` property set by `WebappsConfigurationInitializer`

romansmirnov commented on dist/src/main/java/io/camunda/webapps/controllers/ForwardErrorController.java:
```diff
@@ -0,0 +1,125 @@
+
+  private List<RequestMatcher> ignoredRequestsMatchers;
+
+  @PostConstruct
+  public void init() {
+    // list of endpoints for which 404 is returned
+    ignoredRequestsMatchers =
+        List.of(
+            new AntPathRequestMatcher(PUBLIC_API),
+            new AntPathRequestMatcher(API),
+            new AntPathRequestMatcher(NO_PERMISSION),
+            new AntPathRequestMatcher(SSO_CALLBACK_URI),
+            new AntPathRequestMatcher(IDENTITY_CALLBACK_URI),
+            // actuator endpoints
+            new AntPathRequestMatcher(
+                Optional.ofNullable(environment.getProperty("management.endpoints.web.base-path"))
+                    .map(s -> s.concat("/**"))
+                    .orElse(DEFAULT_MANAGEMENT_URI)));
+  }
```
I need to understand this better, but my first instinct is to not manage such an additional list of paths to exclude them.

romansmirnov reviewed (changes requested):
I have some comments. We have to do a few things differently to avoid having profile checks all over the code base, for example. Or how we configure component scanning, so that there is one way of doing it.

romansmirnov reviewed (approved):
Thanks, @houssain-barouni, that you adjusted the PR. Looks good to me!

* If added a small commit to clean up the `StandaloneTasklist` application.
* As discussed, there is a small bug: When starting `StandaloneTasklist`, the Operate frontend is served when accessing `http://localhost:8080/operate` (and vice versa).

Feel free to merge the PR after fixing that small bug.

vsgoulart commented on tasklist/config/docker-compose.identity.yml:
```diff
@@ -187,7 +187,7 @@ services:
         condition: service_started
     restart: on-failure
-    image: camunda/tasklist:8.5.1
+    image: camunda/tasklist:local-hb
```
@houssain-barouni Should this be left like this?

houssain-barouni replied:
no, it is a local docker image, I will fix it

houssain-barouni commented on tasklist/client/e2e/pages/TestSetupPage.ts:
```diff
@@ -16,7 +16,7 @@ class TestSetupPage {
 
   async goToLoginPage() {
-    await this.page.goto('/login');
+    await this.page.goto('./login');
```
@vsgoulart playwright [docs](https://playwright.dev/docs/api/class-testoptions#test-options-base-url) says about `baseURL`  configuration:

>When using `page.goto()`, `page.route()`, `page.waitForURL()`, `page.waitForRequest()`, or `page.waitForResponse()` it takes the base URL in consideration by using the `URL()` constructor for building the corresponding URL. Unset by default. Examples:
>	•	`baseURL: http://localhost:3000` and navigating to `/bar.html` results in `http://localhost:3000/bar.html`
>	•	`baseURL: `http://localhost:3000/foo/` and navigating to `./bar.html` results in `http://localhost:3000/foo/bar.html`
>	•	`baseURL: http://localhost:3000/foo` (without trailing slash) and navigating to `./bar.html` results in `http://localhost:3000/bar.html`

In my case, I am using in `playwright.config.ts` `baseURL: http://localhost:8081/tasklist/` which matches the second case. I had to add the trailing slash otherwise `expect(page).toHaveURL` was not working.

So as explained by the doc, the `goto` uses `baseURL`  configuration and not the current page to build the target url

houssain-barouni replied:
@vsgoulart took care of updating the paths :)

houssain-barouni commented:
For transparency, a discussion with @vsgoulart about public-forms E2E test that always fail the first time on this branch, but always passes on main:

On `main` branch:
- the process is not found in the first hit when `/new/{processId}` .you can also notice the title is missing
- even though the process is not found, the backend returns `200` and forwards to `index.html`
- the reason is that:
  - `PublicProcessController` does not extend `ApiErrorController` , so all the exceptions are transformed into `500`
  - we had an `ErrorController` that catches all `500` responses and forwards to `index.html `with `200` status
- the UI then fetches the form, which can be not found in the first hit, but the UI does retries to fetch the form
this ends-up by working after some retries

On this branch, as the `ErrorController` was removed, the backend returns `500` when the process is not found

To fix this:
- `PublicProcessController` should extend `ApiErrorController` so that 404 is returned when process is not found
- then we have two options:
  - Make the UI handle the 404 returned `/new/{processId}` , do some retries, and display the error-robot if the process cannot be found
  - catch the `404` in backend and return`index.html`with `200` status, the UI will fail to fetch the form if the process does not really exist (as before)

pedesen reviewed (changes requested):
@houssain-barouni There is an issue with the local dev environment the Operate FE team is using.

- if you run `yarn start` in `operate/client/`, Operate provides a hot reload dev environment running at `localhost:3000`
- with your changes `/client-config.js` can't be retrieved anymore in that dev environment

I think the reason for this is that we use a middleware (see [setupProxy.js](https://github.com/camunda/camunda/blob/main/operate/client/src/setupProxy.js)) to redirect `localhost:3000/client-config.js` to `localhost:8080/client-config.js`. But now it needs to be redirected to `localhost:8080/operate/client-config.ts`

So you would either move client-config.ts or redirect to `localhost:8080/operate/client-config.ts`

houssain-barouni commented:
> @houssain-barouni There is an issue with the local dev environment the Operate FE team is using.
> 
> * if you run `yarn start` in `operate/client/`, Operate provides a hot reload dev environment running at `localhost:3000`
> * with your changes `/client-config.js` can't be retrieved anymore in that dev environment
> 
> I think the reason for this is that we use a middleware (see [setupProxy.js](https://github.com/camunda/camunda/blob/main/operate/client/src/setupProxy.js)) to redirect `localhost:3000/client-config.js` to `localhost:8080/client-config.js`. But now it needs to be redirected to `localhost:8080/operate/client-config.ts`
> 
> So you would either move client-config.ts or redirect to `localhost:8080/operate/client-config.ts`

I added the base path in [setupProxy.js](https://github.com/camunda/camunda/blob/main/operate/client/src/setupProxy.js) and also added `-DskipQA` in makefile commands. QA modules depend on dist so it can fail if `/dist` are not built beforehand, and QA modules are not needed to start the application

pedesen commented on operate/client/src/setupProxy.js:
```diff
@@ -15,7 +15,7 @@ module.exports = function (app) {
     createProxyMiddleware({
       target: `http://localhost:${process.env.IS_E2E ? '8081' : '8080'}`,
-      pathFilter: ['/client-config.js', '/api'],
+      pathFilter: ['/operate/client-config.js', '/api'],
```
> I added the base path in [setupProxy.js](https://github.com/camunda/camunda/blob/main/operate/client/src/setupProxy.js) , and tested that is working with `yarn start` 

How did you test it? For me it doesn't work. The middleware matches the request.url in `pathFilter` and redirects to `target`.

So in this case it would match `localhost:3000/operate/client-config.js` and redirect to `localhost:8080/client-config` which is not what we want. 

What we can do is to add a pathRewrite like this: 
```suggestion
      pathFilter: ['/client-config.js', '/api'],
      pathRewrite: {'^/client-config.js': '/operate/client-config.js'}
```

Then Operate works as expected under `localhost:3000/operate`. This is not ideal because `yarn start` opens the browser at `localhost:3000` which is then empty. But I don't have a solution for this right now.

houssain-barouni replied:
It seems I did not test the right thing,  I just checked `http://localhost:3000/client-config.js` in the Dev console, and it seems that it was just rerouted to index

houssain-barouni replied:
@pedesen I solved the `localhost:3000` by adding `/` in `pathFilter`
localhost:3000 -> localhost:8080 -> localhost:8080/operate (the last redirection is done by backend)

pedesen replied:
it works, but then frontend code from the maven build is used (served at port 8080). What we need is the frontend code served by CRA at 3000, so that hot reload still works. So ideally a redirect `localhost:3000/operate` -> `localhost:3000`, but not sure how this is possible using the middleware, because it's a different target port :thinking:

houssain-barouni replied:
I will remove the `/` from the filter

houssain-barouni replied:
done

pedesen reviewed (changes requested):
I checked that everything works fine. There is just one change needed in the playwright.config (see my comment there).

houssain-barouni commented on operate/client/e2e-playwright/tests/login.spec.ts:
```diff
@@ -37,7 +37,7 @@ test.describe('login page', () => {
     });
 
-    await expect(page).toHaveURL(Paths.dashboard());
+    await expect(page).toHaveURL('../operate'); // dashboard url, we need to do this because baseURL contains a slash at the end as expected by playwright
```
@pedesen maybe we can use absolute path like @vsgoulart did here https://github.com/camunda/camunda/pull/18604/commits/9ec027e00fe38bcc55655071939566db3ddd4908

pedesen replied:
Done!

pedesen replied:
I realized we're reusing the navigate functions in a11y and visual regression tests, so we can't add absolute paths there. I just reverted my commit.

houssain-barouni commented on operate/client/e2e-playwright/test-fixtures.ts:
```diff
@@ -68,6 +68,13 @@ const test = base.extend<
+
       // Important: make sure we authenticate in a clean environment by unsetting storage state.
       const page = await browser.newPage({storageState: undefined});
       await page.goto(`${baseURL}/login`);
```
Hi @pedesen 
`new URL('/login', baseURL).toString()` uses absolute path would return `http://localhost:8080/login` instead of `http://localhost:8080/operate/login`
I can use absolute path `new URL('/operate/login', baseURL).toString()` but that would make the other CIs failing

pedesen replied:
For transparency: We discussed to leave it as it is, because it was not possible to use `new URL` without breaking other things.
